crates/circuit/src/dag_node.rs

```
@@ -76,114 +109,121 @@ pub struct DAGOpNode {
+            py_op: RefCell::new(Some(op.unbind())),
+        };
+
+        Py::new(
+            py,
+            (
+                DAGOpNode {
+                    instruction,
+                    sort_key,
+                },
+                DAGNode { node: None },
+            ),
+        )
+    }
+
+    fn __hash__(slf: PyRef<'_, Self>) -> PyResult<u64> {
+        let super_ = slf.as_ref();
+        let mut hasher = AHasher::default();
+        hasher.write_isize(super_.py_nid());
+        hasher.write(slf.instruction.operation.name().as_bytes());
+        Ok(hasher.finish())
+    }
+
+    fn __eq__(slf: PyRef<Self>, py: Python, other: &Bound<PyAny>) -> PyResult<bool> {
+        let Ok(other) = other.downcast::<Self>() else {
+            return Ok(false);
+        };
+        let borrowed_other = other.borrow();
+        let other_super = borrowed_other.as_ref();
+        let super_ = slf.as_ref();
+
+        if super_.py_nid() != other_super.py_nid() {
+            return Ok(false);
+        }
+        if !slf
+            .instruction
+            .operation
+            .py_eq(py, &borrowed_other.instruction.operation)?
+        {
+            return Ok(false);
+        }
+        let params_eq = if slf.instruction.operation.try_standard_gate().is_some() {
+            slf.instruction
+                .params
+                .iter()
+                .zip(borrowed_other.instruction.params.iter())
+                .all(|(a, b)| match [a, b] {
+                    [Param::Float(float_a), Param::Float(float_b)] => {
+                        relative_eq!(float_a, float_b, max_relative = 1e-10)
                     }
-                }
-            }
-            None => qargs.str()?.into_any(),
+                    [Param::ParameterExpression(param_a), Param::ParameterExpression(param_b)] => {
+                        param_a.bind(py).eq(param_b).unwrap()
+                    }
+                    _ => false,
+                })
```

This doesn't feel quite right for the `Param` comparison, since floats can compare equal to things in `ParameterExpression` (fully bound instances) and `Obj` (integers) variants of `Param`.

Maybe making a `Param::py_eq_approx` or something that takes the `max_relative` as a parameter, and handles all branches with proper equality checks (like `Param::eq` does) might be safer?

---

I did this for semantic equality but I think we're ok here though because the `DAGOpNode.__eq__` method here is really supposed to be used in place of the object identity and set/dict membership which is no longer reliable with this PR. I opted to keep this stricter so we're not accidentally treating equivalent op nodes as the same.

---

Ok, I'm fine with that in principle.  Can you add a comment explaining it just above?

I think we still might need to accept `[Param::Obj(a), Param::Obj(b)] => a.bind(py).eq(b)` in that form though.

I don't feel great about suppressing the exceptions in the `ParameterExpression` case (nor the proposed `Obj` case) - can we propagate that up?

---

Updated in: https://github.com/Qiskit/qiskit/pull/12550/commits/701c980e120221bc968b76a721227f82ecc6007f